File: mobile-sdk/flutter-sdk/lib/offline/offline_manager.dart
import 'dart:async';
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:connectivity_plus/connectivity_plus.dart';
import 'package:rxdart/rxdart.dart';
import 'package:uuid/uuid.dart';

import '../api/api_client.dart';
import '../types/types.dart';
import '../utils/constants.dart';
import '../utils/retry.dart';

class OfflineManagerConfig {
  final int syncInterval;
  final int maxRetryAttempts;
  final bool enableOptimisticUpdates;
  final bool enableBackgroundSync;

  const OfflineManagerConfig({
    this.syncInterval = SyncIntervals.normal,
    this.maxRetryAttempts = 3,
    this.enableOptimisticUpdates = true,
    this.enableBackgroundSync = true,
  });
}

class OfflineManager {
  final ApiClient apiClient;
  final OfflineManagerConfig config;
  final Connectivity _connectivity;
  final Uuid _uuid;
  
  Timer? _syncTimer;
  bool _isOnline = true;
  bool _isSyncing = false;
  
  Map<String, List<Record>> _cachedTables = {};
  List<SyncOperation> _pendingOperations = [];
  int _lastSync = 0;
  
  // Event streams
  final _networkController = StreamController<bool>.broadcast();
  final _syncStartController = StreamController<void>.broadcast();
  final _syncCompleteController = StreamController<void>.broadcast();
  final _syncErrorController = StreamController<PyAirtableException>.broadcast();
  
  Stream<bool> get onNetworkChange => _networkController.stream;
  Stream<void> get onSyncStart => _syncStartController.stream;
  Stream<void> get onSyncComplete => _syncCompleteController.stream;
  Stream<PyAirtableException> get onSyncError => _syncErrorController.stream;

  OfflineManager({
    required this.apiClient,
    required this.config,
  }) : _connectivity = Connectivity(),
       _uuid = const Uuid();

  Future<void> initialize() async {
    await _loadOfflineData();
    _setupNetworkListener();
    
    if (config.enableBackgroundSync) {
      _startSyncTimer();
    }
  }

  void _setupNetworkListener() {
    _connectivity.onConnectivityChanged.listen((result) {
      final wasOnline = _isOnline;
      _isOnline = result != ConnectivityResult.none;
      
      _networkController.add(_isOnline);
      
      // When coming back online, trigger sync
      if (!wasOnline && _isOnline) {
        _debouncedSync();
      }
    });
  }

  Future<List<Record>> getCachedRecords(String baseId, String tableId) async {
    final tableKey = '$baseId:$tableId';
    return _cachedTables[tableKey] ?? [];
  }

  Future<void> cacheRecords(String baseId, String tableId, List<Record> records) async {
    final tableKey = '$baseId:$tableId';
    _cachedTables[tableKey] = records;
    await _saveOfflineData();
  }

  Future<Record> createRecord(String baseId, String tableId, RecordCreate record) async {
    final operation = SyncOperation(
      id: _uuid.v4(),
      type: 'create',
      tableName: '$baseId:$tableId',
      data: record.toJson(),
      timestamp: DateTime.now().millisecondsSinceEpoch,
      status: 'pending',
      retryCount: 0,
    );

    // If online, try immediate sync
    if (_isOnline) {
      try {
        final result = await apiClient.createRecord(baseId, tableId, record);
        await _updateRecordInCache(baseId, tableId, result);
        return result;
      } catch (error) {
        // If failed and optimistic updates enabled, queue for later
        if (config.enableOptimisticUpdates) {
          return _createOptimisticRecord(baseId, tableId, record, operation);
        }
        rethrow;
      }
    }

    // Offline: create optimistic record and queue operation
    if (config.enableOptimisticUpdates) {
      return _createOptimisticRecord(baseId, tableId, record, operation);
    }

    throw const SyncException('Cannot create record while offline');
  }

  Future<Record> updateRecord(String baseId, String tableId, RecordUpdate record) async {
    final operation = SyncOperation(
      id: _uuid.v4(),
      type: 'update',
      tableName: '$baseId:$tableId',
      recordId: record.id,
      data: record.toJson(),
      timestamp: DateTime.now().millisecondsSinceEpoch,
      status: 'pending',
      retryCount: 0,
    );

    // If online, try immediate sync
    if (_isOnline) {
      try {
        final result = await apiClient.updateRecord(baseId, tableId, record);
        await _updateRecordInCache(baseId, tableId, result);
        return result;
      } catch (error) {
        // If failed and optimistic updates enabled, queue for later
        if (config.enableOptimisticUpdates) {
          return _updateOptimisticRecord(baseId, tableId, record, operation);
        }
        rethrow;
      }
    }

    // Offline: create optimistic update and queue operation
    if (config.enableOptimisticUpdates) {
      return _updateOptimisticRecord(baseId, tableId, record, operation);
    }

    throw const SyncException('Cannot update record while offline');
  }

  Future<void> deleteRecord(String baseId, String tableId, String recordId) async {
    final operation = SyncOperation(
      id: _uuid.v4(),
      type: 'delete',
      tableName: '$baseId:$tableId',
      recordId: recordId,
      data: {'id': recordId},
      timestamp: DateTime.now().millisecondsSinceEpoch,
      status: 'pending',
      retryCount: 0,
    );

    // If online, try immediate sync
    if (_isOnline) {
      try {
        await apiClient.deleteRecord(baseId, tableId, recordId);
        await _removeRecordFromCache(baseId, tableId, recordId);
        return;
      } catch (error) {
        // If failed and optimistic updates enabled, queue for later
        if (config.enableOptimisticUpdates) {
          await _deleteOptimisticRecord(baseId, tableId, recordId, operation);
          return;
        }
        rethrow;
      }
    }

    // Offline: create optimistic delete and queue operation
    if (config.enableOptimisticUpdates) {
      await _deleteOptimisticRecord(baseId, tableId, recordId, operation);
      return;
    }

    throw const SyncException('Cannot delete record while offline');
  }

  Future<void> sync() async {
    if (_isSyncing || !_isOnline) return;

    _isSyncing = true;
    _syncStartController.add(null);

    try {
      await _processPendingOperations();
      _lastSync = DateTime.now().millisecondsSinceEpoch;
      await _saveOfflineData();
      
      _syncCompleteController.add(null);
    } catch (error) {
      final exception = error is PyAirtableException
          ? error
          : PyAirtableException(error.toString());
      _syncErrorController.add(exception);
      rethrow;
    } finally {
      _isSyncing = false;
    }
  }

  int get pendingOperationsCount => _pendingOperations.length;

  List<SyncOperation> get pendingOperations => List.unmodifiable(_pendingOperations);

  Future<void> clearCache() async {
    _cachedTables.clear();
    _pendingOperations.clear();
    _lastSync = 0;
    await _saveOfflineData();
  }

  // Private methods

  Future<Record> _createOptimisticRecord(
    String baseId,
    String tableId,
    RecordCreate record,
    SyncOperation operation,
  ) async {
    final optimisticRecord = Record(
      id: 'temp_${_uuid.v4()}',
      fields: record.fields,
      createdTime: DateTime.now().toIso8601String(),
      version: 0,
    );

    await _updateRecordInCache(baseId, tableId, optimisticRecord);
    _pendingOperations.add(operation);
    await _saveOfflineData();

    return optimisticRecord;
  }

  Future<Record> _updateOptimisticRecord(
    String baseId,
    String tableId,
    RecordUpdate record,
    SyncOperation operation,
  ) async {
    final existing = await _getRecordFromCache(baseId, tableId, record.id);
    
    final updatedRecord = Record(
      id: record.id,
      fields: record.fields,
      createdTime: existing?.createdTime ?? DateTime.now().toIso8601String(),
      updatedTime: DateTime.now().toIso8601String(),
      version: (existing?.version ?? 0) + 1,
    );

    await _updateRecordInCache(baseId, tableId, updatedRecord);
    _pendingOperations.add(operation);
    await _saveOfflineData();

    return updatedRecord;
  }

  Future<void> _deleteOptimisticRecord(
    String baseId,
    String tableId,
    String recordId,
    SyncOperation operation,
  ) async {
    await _removeRecordFromCache(baseId, tableId, recordId);
    _pendingOperations.add(operation);
    await _saveOfflineData();
  }

  Future<void> _processPendingOperations() async {
    final operations = List<SyncOperation>.from(_pendingOperations);
    
    for (final operation in operations) {
      try {
        final updatedOperation = operation.copyWith(status: 'syncing');
        await _processOperation(updatedOperation);
        
        // Remove from pending operations
        _pendingOperations.removeWhere((op) => op.id == operation.id);
      } catch (error) {
        final updatedOperation = operation.copyWith(
          status: 'failed',
          retryCount: operation.retryCount + 1,
        );
        
        if (updatedOperation.retryCount >= config.maxRetryAttempts) {
          // Remove failed operation after max retries
          _pendingOperations.removeWhere((op) => op.id == operation.id);
        } else {
          // Update operation with new retry count
          final index = _pendingOperations.indexWhere((op) => op.id == operation.id);
          if (index >= 0) {
            _pendingOperations[index] = updatedOperation;
          }
        }
      }
    }
  }

  Future<void> _processOperation(SyncOperation operation) async {
    final parts = operation.tableName.split(':');
    final baseId = parts[0];
    final tableId = parts[1];
    
    switch (operation.type) {
      case 'create':
        final recordCreate = RecordCreate(fields: operation.data['fields']);
        await apiClient.createRecord(baseId, tableId, recordCreate);
        break;
        
      case 'update':
        final recordUpdate = RecordUpdate(
          id: operation.recordId!,
          fields: operation.data['fields'],
        );
        await apiClient.updateRecord(baseId, tableId, recordUpdate);
        break;
        
      case 'delete':
        await apiClient.deleteRecord(baseId, tableId, operation.recordId!);
        break;
    }
  }

  Future<void> _updateRecordInCache(String baseId, String tableId, Record record) async {
    final tableKey = '$baseId:$tableId';
    final records = _cachedTables[tableKey] ?? <Record>[];
    
    final existingIndex = records.indexWhere((r) => r.id == record.id);
    if (existingIndex >= 0) {
      records[existingIndex] = record;
    } else {
      records.add(record);
    }
    
    _cachedTables[tableKey] = records;
    await _saveOfflineData();
  }

  Future<Record?> _getRecordFromCache(String baseId, String tableId, String recordId) async {
    final tableKey = '$baseId:$tableId';
    final records = _cachedTables[tableKey] ?? <Record>[];
    try {
      return records.firstWhere((r) => r.id == recordId);
    } catch (e) {
      return null;
    }
  }

  Future<void> _removeRecordFromCache(String baseId, String tableId, String recordId) async {
    final tableKey = '$baseId:$tableId';
    final records = _cachedTables[tableKey] ?? <Record>[];
    
    records.removeWhere((r) => r.id == recordId);
    _cachedTables[tableKey] = records;
    await _saveOfflineData();
  }

  Future<void> _loadOfflineData() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final data = prefs.getString(StorageKeys.offlineData);
      
      if (data != null) {
        final jsonData = jsonDecode(data) as Map<String, dynamic>;
        
        // Load cached tables
        final tablesData = jsonData['tables'] as Map<String, dynamic>? ?? {};
        _cachedTables = tablesData.map((key, value) => MapEntry(
          key,
          (value as List).map((record) => Record.fromJson(record)).toList(),
        ));
        
        // Load pending operations
        final operationsData = jsonData['pendingOperations'] as List? ?? [];
        _pendingOperations = operationsData
            .map((op) => SyncOperation.fromJson(op))
            .toList();
        
        _lastSync = jsonData['lastSync'] as int? ?? 0;
      }
    } catch (error) {
      print('Failed to load offline data: $error');
    }
  }

  Future<void> _saveOfflineData() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      
      final data = {
        'tables': _cachedTables.map((key, value) =>
            MapEntry(key, value.map((record) => record.toJson()).toList())),
        'pendingOperations': _pendingOperations.map((op) => op.toJson()).toList(),
        'lastSync': _lastSync,
      };
      
      await prefs.setString(StorageKeys.offlineData, jsonEncode(data));
    } catch (error) {
      print('Failed to save offline data: $error');
    }
  }

  void _startSyncTimer() {
    _syncTimer = Timer.periodic(
      Duration(milliseconds: config.syncInterval),
      (_) {
        if (_isOnline && !_isSyncing && _pendingOperations.isNotEmpty) {
          _debouncedSync();
        }
      },
    );
  }

  late final Debouncer _syncDebouncer = Debouncer(milliseconds: 1000);
  
  void _debouncedSync() {
    _syncDebouncer.run(() {
      sync().catchError((error) {
        print('Background sync failed: $error');
      });
    });
  }

  Future<void> dispose() async {
    _syncTimer?.cancel();
    _syncDebouncer.dispose();
    
    await _networkController.close();
    await _syncStartController.close();
    await _syncCompleteController.close();
    await _syncErrorController.close();
  }
}

File: mobile-sdk/react-native-sdk/src/offline/OfflineManager.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import NetInfo from '@react-native-community/netinfo';
import { EventEmitter } from 'events';
import { 
  Record, 
  RecordCreate, 
  RecordUpdate, 
  SyncOperation, 
  OfflineData,
  ApiResponse 
} from '../types';
import { ApiClient } from '../api/ApiClient';
import { STORAGE_KEYS, SYNC_INTERVALS } from '../utils/constants';
import { SyncError, PyAirtableSDKError } from '../utils/errors';
import { debounce } from '../utils/retry';
import uuid from 'react-native-uuid';

export interface OfflineManagerOptions {
  syncInterval?: number;
  maxRetryAttempts?: number;
  enableOptimisticUpdates?: boolean;
  enableBackgroundSync?: boolean;
}

export class OfflineManager extends EventEmitter {
  private apiClient: ApiClient;
  private options: Required<OfflineManagerOptions>;
  private syncTimer: NodeJS.Timeout | null = null;
  private isOnline = true;
  private isSyncing = false;
  private offlineData: OfflineData = {
    tables: {},
    lastSync: 0,
    pendingOperations: [],
  };

  constructor(apiClient: ApiClient, options: OfflineManagerOptions = {}) {
    super();
    this.apiClient = apiClient;
    this.options = {
      syncInterval: SYNC_INTERVALS.NORMAL,
      maxRetryAttempts: 3,
      enableOptimisticUpdates: true,
      enableBackgroundSync: true,
      ...options,
    };

    this.initialize();
  }

  /**
   * Initialize offline manager
   */
  private async initialize(): Promise<void> {
    await this.loadOfflineData();
    this.setupNetworkListener();
    
    if (this.options.enableBackgroundSync) {
      this.startSyncTimer();
    }
  }

  /**
   * Setup network connectivity listener
   */
  private setupNetworkListener(): void {
    NetInfo.addEventListener(state => {
      const wasOnline = this.isOnline;
      this.isOnline = state.isConnected ?? false;
      
      // When coming back online, trigger sync
      if (!wasOnline && this.isOnline) {
        this.emit('online');
        this.debouncedSync();
      } else if (wasOnline && !this.isOnline) {
        this.emit('offline');
      }
    });
  }

  /**
   * Get cached records for a table
   */
  async getCachedRecords(baseId: string, tableId: string): Promise<Record[]> {
    const tableKey = `${baseId}:${tableId}`;
    return this.offlineData.tables[tableKey] || [];
  }

  /**
   * Cache records for a table
   */
  async cacheRecords(baseId: string, tableId: string, records: Record[]): Promise<void> {
    const tableKey = `${baseId}:${tableId}`;
    this.offlineData.tables[tableKey] = records;
    await this.saveOfflineData();
    this.emit('cacheUpdated', { baseId, tableId, records });
  }

  /**
   * Create record (with offline support)
   */
  async createRecord(
    baseId: string, 
    tableId: string, 
    record: RecordCreate
  ): Promise<Record> {
    const operation: SyncOperation = {
      id: uuid.v4() as string,
      type: 'create',
      tableName: `${baseId}:${tableId}`,
      data: record,
      timestamp: Date.now(),
      status: 'pending',
      retryCount: 0,
    };

    // If online, try immediate sync
    if (this.isOnline) {
      try {
        const result = await this.apiClient.createRecord(baseId, tableId, record);
        
        // Update cache with created record
        await this.updateRecordInCache(baseId, tableId, result);
        
        return result;
      } catch (error) {
        // If failed and optimistic updates enabled, queue for later
        if (this.options.enableOptimisticUpdates) {
          return this.createOptimisticRecord(baseId, tableId, record, operation);
        }
        throw error;
      }
    }

    // Offline: create optimistic record and queue operation
    if (this.options.enableOptimisticUpdates) {
      return this.createOptimisticRecord(baseId, tableId, record, operation);
    }

    throw new SyncError('Cannot create record while offline');
  }

  /**
   * Update record (with offline support)
   */
  async updateRecord(
    baseId: string, 
    tableId: string, 
    record: RecordUpdate
  ): Promise<Record> {
    const operation: SyncOperation = {
      id: uuid.v4() as string,
      type: 'update',
      tableName: `${baseId}:${tableId}`,
      recordId: record.id,
      data: record,
      timestamp: Date.now(),
      status: 'pending',
      retryCount: 0,
    };

    // If online, try immediate sync
    if (this.isOnline) {
      try {
        const result = await this.apiClient.updateRecord(baseId, tableId, record);
        
        // Update cache
        await this.updateRecordInCache(baseId, tableId, result);
        
        return result;
      } catch (error) {
        // If failed and optimistic updates enabled, queue for later
        if (this.options.enableOptimisticUpdates) {
          return this.updateOptimisticRecord(baseId, tableId, record, operation);
        }
        throw error;
      }
    }

    // Offline: create optimistic update and queue operation
    if (this.options.enableOptimisticUpdates) {
      return this.updateOptimisticRecord(baseId, tableId, record, operation);
    }

    throw new SyncError('Cannot update record while offline');
  }

  /**
   * Delete record (with offline support)
   */
  async deleteRecord(baseId: string, tableId: string, recordId: string): Promise<void> {
    const operation: SyncOperation = {
      id: uuid.v4() as string,
      type: 'delete',
      tableName: `${baseId}:${tableId}`,
      recordId,
      data: { id: recordId },
      timestamp: Date.now(),
      status: 'pending',
      retryCount: 0,
    };

    // If online, try immediate sync
    if (this.isOnline) {
      try {
        await this.apiClient.deleteRecord(baseId, tableId, recordId);
        
        // Remove from cache
        await this.removeRecordFromCache(baseId, tableId, recordId);
        
        return;
      } catch (error) {
        // If failed and optimistic updates enabled, queue for later
        if (this.options.enableOptimisticUpdates) {
          await this.deleteOptimisticRecord(baseId, tableId, recordId, operation);
          return;
        }
        throw error;
      }
    }

    // Offline: create optimistic delete and queue operation
    if (this.options.enableOptimisticUpdates) {
      await this.deleteOptimisticRecord(baseId, tableId, recordId, operation);
      return;
    }

    throw new SyncError('Cannot delete record while offline');
  }

  /**
   * Force sync with server
   */
  async sync(): Promise<void> {
    if (this.isSyncing || !this.isOnline) {
      return;
    }

    this.isSyncing = true;
    this.emit('syncStart');

    try {
      await this.processPendingOperations();
      this.offlineData.lastSync = Date.now();
      await this.saveOfflineData();
      
      this.emit('syncComplete');
    } catch (error) {
      this.emit('syncError', error);
      throw error;
    } finally {
      this.isSyncing = false;
    }
  }

  /**
   * Get pending operations count
   */
  getPendingOperationsCount(): number {
    return this.offlineData.pendingOperations.length;
  }

  /**
   * Get pending operations
   */
  getPendingOperations(): SyncOperation[] {
    return [...this.offlineData.pendingOperations];
  }

  /**
   * Clear all cached data
   */
  async clearCache(): Promise<void> {
    this.offlineData = {
      tables: {},
      lastSync: 0,
      pendingOperations: [],
    };
    await this.saveOfflineData();
    this.emit('cacheCleared');
  }

  /**
   * Create optimistic record
   */
  private async createOptimisticRecord(
    baseId: string,
    tableId: string,
    record: RecordCreate,
    operation: SyncOperation
  ): Promise<Record> {
    const optimisticRecord: Record = {
      id: `temp_${uuid.v4()}`,
      fields: record.fields,
      createdTime: new Date().toISOString(),
      version: 0,
    };

    // Add to cache
    await this.updateRecordInCache(baseId, tableId, optimisticRecord);
    
    // Queue operation
    this.offlineData.pendingOperations.push(operation);
    await this.saveOfflineData();

    return optimisticRecord;
  }

  /**
   * Update optimistic record
   */
  private async updateOptimisticRecord(
    baseId: string,
    tableId: string,
    record: RecordUpdate,
    operation: SyncOperation
  ): Promise<Record> {
    const updatedRecord: Record = {
      id: record.id,
      fields: record.fields,
      createdTime: '', // Will be filled from cache
      updatedTime: new Date().toISOString(),
      version: 0,
    };

    // Get existing record from cache to preserve createdTime
    const existing = await this.getRecordFromCache(baseId, tableId, record.id);
    if (existing) {
      updatedRecord.createdTime = existing.createdTime;
      updatedRecord.version = (existing.version || 0) + 1;
    }

    // Update cache
    await this.updateRecordInCache(baseId, tableId, updatedRecord);
    
    // Queue operation
    this.offlineData.pendingOperations.push(operation);
    await this.saveOfflineData();

    return updatedRecord;
  }

  /**
   * Delete optimistic record
   */
  private async deleteOptimisticRecord(
    baseId: string,
    tableId: string,
    recordId: string,
    operation: SyncOperation
  ): Promise<void> {
    // Remove from cache
    await this.removeRecordFromCache(baseId, tableId, recordId);
    
    // Queue operation
    this.offlineData.pendingOperations.push(operation);
    await this.saveOfflineData();
  }

  /**
   * Process pending sync operations
   */
  private async processPendingOperations(): Promise<void> {
    const operations = [...this.offlineData.pendingOperations];
    
    for (const operation of operations) {
      try {
        operation.status = 'syncing';
        await this.processOperation(operation);
        
        // Remove from pending operations
        this.offlineData.pendingOperations = this.offlineData.pendingOperations
          .filter(op => op.id !== operation.id);
          
      } catch (error) {
        operation.status = 'failed';
        operation.retryCount++;
        
        if (operation.retryCount >= this.options.maxRetryAttempts) {
          // Remove failed operation after max retries
          this.offlineData.pendingOperations = this.offlineData.pendingOperations
            .filter(op => op.id !== operation.id);
          
          this.emit('operationFailed', { operation, error });
        }
      }
    }
  }

  /**
   * Process a single sync operation
   */
  private async processOperation(operation: SyncOperation): Promise<void> {
    const [baseId, tableId] = operation.tableName.split(':');
    
    switch (operation.type) {
      case 'create':
        await this.apiClient.createRecord(baseId, tableId, operation.data);
        break;
        
      case 'update':
        await this.apiClient.updateRecord(baseId, tableId, operation.data);
        break;
        
      case 'delete':
        await this.apiClient.deleteRecord(baseId, tableId, operation.recordId!);
        break;
    }
  }

  /**
   * Update record in cache
   */
  private async updateRecordInCache(
    baseId: string, 
    tableId: string, 
    record: Record
  ): Promise<void> {
    const tableKey = `${baseId}:${tableId}`;
    const records = this.offlineData.tables[tableKey] || [];
    
    const existingIndex = records.findIndex(r => r.id === record.id);
    if (existingIndex >= 0) {
      records[existingIndex] = record;
    } else {
      records.push(record);
    }
    
    this.offlineData.tables[tableKey] = records;
    await this.saveOfflineData();
  }

  /**
   * Get record from cache
   */
  private async getRecordFromCache(
    baseId: string, 
    tableId: string, 
    recordId: string
  ): Promise<Record | null> {
    const tableKey = `${baseId}:${tableId}`;
    const records = this.offlineData.tables[tableKey] || [];
    return records.find(r => r.id === recordId) || null;
  }

  /**
   * Remove record from cache
   */
  private async removeRecordFromCache(
    baseId: string, 
    tableId: string, 
    recordId: string
  ): Promise<void> {
    const tableKey = `${baseId}:${tableId}`;
    const records = this.offlineData.tables[tableKey] || [];
    
    this.offlineData.tables[tableKey] = records.filter(r => r.id !== recordId);
    await this.saveOfflineData();
  }

  /**
   * Load offline data from storage
   */
  private async loadOfflineData(): Promise<void> {
    try {
      const data = await AsyncStorage.getItem(STORAGE_KEYS.OFFLINE_DATA);
      if (data) {
        this.offlineData = JSON.parse(data);
      }
    } catch (error) {
      console.warn('Failed to load offline data:', error);
    }
  }

  /**
   * Save offline data to storage
   */
  private async saveOfflineData(): Promise<void> {
    try {
      await AsyncStorage.setItem(
        STORAGE_KEYS.OFFLINE_DATA, 
        JSON.stringify(this.offlineData)
      );
    } catch (error) {
      console.warn('Failed to save offline data:', error);
    }
  }

  /**
   * Start sync timer
   */
  private startSyncTimer(): void {
    this.syncTimer = setInterval(() => {
      if (this.isOnline && !this.isSyncing && this.offlineData.pendingOperations.length > 0) {
        this.debouncedSync();
      }
    }, this.options.syncInterval);
  }

  /**
   * Debounced sync to prevent too frequent syncs
   */
  private debouncedSync = debounce(() => {
    this.sync().catch(error => {
      console.warn('Background sync failed:', error);
    });
  }, 1000);

  /**
   * Stop sync timer and cleanup
   */
  destroy(): void {
    if (this.syncTimer) {
      clearInterval(this.syncTimer);
      this.syncTimer = null;
    }
    
    this.removeAllListeners();
  }
}